### ISSUE TITLE
skip foreign key checks

### ISSUE THREAD

**Issue by babinomec** [github/gh-ost#312]
It would be nice to have an option like "--trust-me-i-there-are-no-foreign-keys-for-this-table" in order to save some time (4/5 minutes in our case, we have around 300k tables)

**Comment by ggunson** [comment]
The query for getting the foreign keys is this:

```sql
		SELECT
			SUM(REFERENCED_TABLE_NAME IS NOT NULL AND TABLE_SCHEMA=? AND TABLE_NAME=?) as num_child_side_fk,
			SUM(REFERENCED_TABLE_NAME IS NOT NULL AND REFERENCED_TABLE_SCHEMA=? AND REFERENCED_TABLE_NAME=?) as num_parent_side_fk
		FROM INFORMATION_SCHEMA.KEY_COLUMN_USAGE
		WHERE
				REFERENCED_TABLE_NAME IS NOT NULL
				AND ((TABLE_SCHEMA=? AND TABLE_NAME=?)
					OR (REFERENCED_TABLE_SCHEMA=? AND REFERENCED_TABLE_NAME=?)
				)
```

The EXPLAIN:

```
*************************** 1. row ***************************
           id: 1
  select_type: SIMPLE
        table: KEY_COLUMN_USAGE
         type: ALL
possible_keys: NULL
          key: NULL
      key_len: NULL
          ref: NULL
         rows: NULL
        Extra: Using where; Open_full_table; Scanned all databases
1 row in set (0.00 sec)
```

Putting `REFERENCED_TABLE_NAME IS NOT NULL` in the sums is redundant since the check is already in the WHERE clause. I'm curious if changing this into a query with UNION might be faster. 

@babinomec Would you mind trying out these two queries a couple times each on a replica and let me know if the second one is any faster (changing the xxx_data and tttttttt to whatever the original names were):

```sql
		SELECT
			SUM(TABLE_SCHEMA='xxx_data' AND TABLE_NAME='tttttttt') as num_child_side_fk,
			SUM(REFERENCED_TABLE_SCHEMA='xxx_data' AND REFERENCED_TABLE_NAME='tttttttt') as num_parent_side_fk
		FROM INFORMATION_SCHEMA.KEY_COLUMN_USAGE
		WHERE
				REFERENCED_TABLE_NAME IS NOT NULL
				AND ((TABLE_SCHEMA='xxx_data' AND TABLE_NAME='tttttttt')
					OR (REFERENCED_TABLE_SCHEMA='xxx_data' AND REFERENCED_TABLE_NAME='tttttttt')
				);
```

```sql
SELECT SUM(num_child_side_fk), SUM(num_parent_side_fk) FROM
	(	SELECT
			1 as num_child_side_fk,
			0 as num_parent_side_fk
		FROM INFORMATION_SCHEMA.KEY_COLUMN_USAGE
		WHERE
				REFERENCED_TABLE_NAME IS NOT NULL
				AND (TABLE_SCHEMA='xxx_data' AND TABLE_NAME='tttttttt')
		UNION ALL
		SELECT
			0 as num_child_side_fk,
			1 as num_parent_side_fk
		FROM INFORMATION_SCHEMA.KEY_COLUMN_USAGE
		WHERE
				REFERENCED_TABLE_SCHEMA='xxx_data' AND REFERENCED_TABLE_NAME='tttttttt') subq;
```

A flag to ignore foreign keys might be a tad dangerous if you've got `ON DELETE CASCADE`; it definitely wouldn't default to 1. @shlomi-noach did some work on a flag to discard FKs here: https://github.com/github/gh-ost/pull/264

(I originally put this comment in the wrong issue: https://github.com/github/gh-ost/issues/313)

**Comment by babinomec** [comment]
It's pretty much the same (except the first run obviously).

Here are the results for the first query : 
```
+-------------------+--------------------+
| num_child_side_fk | num_parent_side_fk |
+-------------------+--------------------+
|              NULL |               NULL |
+-------------------+--------------------+
1 row in set, 319 warnings (1 min 9.52 sec)

+-------------------+--------------------+
| num_child_side_fk | num_parent_side_fk |
+-------------------+--------------------+
|              NULL |               NULL |
+-------------------+--------------------+
1 row in set, 1 warning (49.44 sec)

+-------------------+--------------------+
| num_child_side_fk | num_parent_side_fk |
+-------------------+--------------------+
|              NULL |               NULL |
+-------------------+--------------------+
1 row in set (47.68 sec)
```

And the second one : 
```
+------------------------+-------------------------+
| SUM(num_child_side_fk) | SUM(num_parent_side_fk) |
+------------------------+-------------------------+
|                   NULL |                    NULL |
+------------------------+-------------------------+
1 row in set, 364 warnings (56.41 sec)

+------------------------+-------------------------+
| SUM(num_child_side_fk) | SUM(num_parent_side_fk) |
+------------------------+-------------------------+
|                   NULL |                    NULL |
+------------------------+-------------------------+
1 row in set, 4 warnings (50.29 sec)

+------------------------+-------------------------+
| SUM(num_child_side_fk) | SUM(num_parent_side_fk) |
+------------------------+-------------------------+
|                   NULL |                    NULL |
+------------------------+-------------------------+
1 row in set, 6 warnings (45.25 sec)
```

**Comment by ggunson** [comment]
@babinomec Thanks for testing that. 

@shlomi-noach What do you think of the --trust-me-i-there-are-no-foreign-keys-for-this-table idea?

**Comment by shlomi-noach** [comment]
Thank you both -- I understand the need; I think it makes sense.

**Comment by babinomec** [comment]
(seems I forgot a "know" in the option name)

**Comment by shlomi-noach** [comment]
The race for option name suggestions is now open

**Comment by ggunson** [comment]
--skip-foreign-key-check 
--ignore-foreign-keys
--trust-me-i-know-what-im-doing
--these-arent-the-droids-youre-looking-for
--pay-no-attention-to-the-man-behind-the-curtain

Would there be any benefit to, instead of just not doing the foreign keys query with this option on, forking the query (as we do with the `SELECT count(*) from table`) and then writing a warning to the output if there are FKs (rather than failing gh-ost on the results being non-null)?

**Comment by shlomi-noach** [comment]
> Would there be any benefit to, instead of just not doing the foreign keys query with this option on, forking the query (as we do with the SELECT count(*) from table) and then writing a warning to the output if there are FKs (rather than failing gh-ost on the results being non-null)? 

@ggunson I think in this case it wouldn't be beneficial, and possible inconsistent. The check itself may terminate at _some_ time during migration, at which time no one pays notice. But for smaller tables, it would not even complete by the time the migration completes.
Given that the user provides a flag `--skip-foreign-key-checks` I would have `gh-ost` actually skip the check.

**Comment by shlomi-noach** [comment]
Please see https://github.com/github/gh-ost/pull/316

**Comment by babinomec** [comment]
Perfect, thank you !

(Am I supposed to close this issue then ?)

**Comment by shlomi-noach** [comment]
@babinomec I'll close it once the PR is merged.

**Comment by shlomi-noach** [comment]
Branch is merged. Closing this issue. There is no release yet that contains this change, next one will.
